sbol_utilities/conversion.py

```diff
@@ -0,0 +1,465 @@
+    :param doc3: SBOL3 document to convert
+    :param path: path to write FASTA file to
+    """
+    with open(path, 'w') as out:
+        for c in id_sort([c for c in doc3.objects if isinstance(c, sbol3.Component)]):
+            # Find all sequences of nucleic acid type
+            na_seqs = [s.lookup() for s in c.sequences if s.lookup().encoding == sbol3.IUPAC_DNA_ENCODING]
+            if len(na_seqs) == 0:  # ignore components with no sequence to serialize
+                continue
+            elif len(na_seqs) == 1:  # if there is precisely one sequence, write it to the FASTA
+                record = SeqIO.SeqRecord(Seq(na_seqs[0].elements), id=c.display_id, description=c.description or '')
+                out.write(record.format('fasta'))
+            else:  # warn about components with ambiguous sequences
+                logging.warning(f'Ambiguous component ({len(na_seqs)} sequences) not converted to FASTA: {c.identity}')
+
+
+def convert_from_fasta(path: str, namespace: str, identity_map: Dict[str, str] = None) -> sbol3.Document:
+    """Convert a FASTA nucleotide document on disk into an SBOL3 document
+    Specifically, every sequence in the FASTA will be converted into an SBOL Component and associated Sequence
+
+    :param path: path to read FASTA file from
+    :param namespace: URIs of Components will be set to {namespace}/{fasta_id}
+    :param identity_map: dictionary mapping from displayId to identity, using for getting non-default namespaces
+    :return: SBOL3 document containing converted materials
+    """
+    doc = sbol3.Document()
+    with open(path, 'r') as f:
+        for r in SeqIO.parse(f, 'fasta'):
+            if identity_map and r.id in identity_map:
+                identity = identity_map[r.id]
+            else:
+                identity = f'{namespace}/{sbol3.string_to_display_id(r.id)}'
+            s = sbol3.Sequence(identity+'_sequence', name=r.name, description=r.description.strip(),
+                               elements=str(r.seq), encoding=sbol3.IUPAC_DNA_ENCODING, namespace=namespace)
+            doc.add(s)
+            doc.add(sbol3.Component(identity, sbol3.SBO_DNA, name=r.name, description=r.description.strip(),
+                                    sequences=[s.identity], namespace=namespace))
+    return doc
+
+
+# TODO: Figure out how to support multiple namespaces like we do for FASTA: currently, importing from multiple namespaces will not work correctly
+def convert_from_genbank(path: str, namespace: str) -> sbol3.Document:
+    """Convert a GenBank document on disk into an SBOL3 document
+    Specifically, the GenBank document is first imported to SBOL2, then converted from SBOL2 to SBOL3
+
+    :param path: path to read GenBank file from
+    :param namespace: URIs of Components will be set to {namespace}/{genbank_id}
+    :return: SBOL3 document containing converted materials
+    """
+    doc2 = sbol2.Document()
+    sbol2.setHomespace(namespace)
+    doc2.importFromFormat(path)
```

Comment:
Eek! This sends the input file over the internet.

You may want to add an argument to `convert_from_genbank` that forces the user to say `online_conversion=True`, and it defaults to `False`, and if it is False, raise an error.

In the future, perhaps we can use BioPython to do local conversion. I don't know how easy or hard that would be.